Refactor remember entitites in shards

johanandren opened the pull request:
Most tests passing, but 4 failing to look into later and a bunch of FIXMEs

References #27763

johanandren committed:
DData and Persistence based remember entitites refactored WIP

Most test passing, but 4 failing to look into later and a bunch of FIXMEs

patriknw reviewed (approved):
Had a brief scan. Good structure and much easier to follow.

patriknw commented:
wasn't intended as approval ;-)

johanandren commented on akka-cluster-sharding/src/test/scala/akka/cluster/sharding/PersistentShardSpec.scala:
```diff
@@ -47,6 +58,7 @@ class PersistentShardSpec extends AkkaSpec(PersistentShardSpec.config) with AnyW
       system.log.info("Starting shard again")
       val secondIncarnation = system.actorOf(props)
 
+      // FIXME how did this ever work when the RestartEntity goes through the parent, expected to be a shard region?
```
Explanation: In PersistentShard/master the GetShardStats is responded with the entity ids from the state even if they are not yet started when the query comes in.

johanandren committed:
PersistentShardSpec fixed

johanandren committed:
Order methods in the order of init in the shard.

Also, skip the await boolean. PersistenceClusterShardingLeaseSpec is now passing.

johanandren commented:
Non multi-jvm tests passing, a few multi-jvm still failing.

johanandren committed:
Multi JVM tests passing but SupervisionSpec failing

johanandren committed:
Some bad isolation between test cases causing problems

johanandren committed:
Test coverage for remember entities store failures

johanandren committed:
Lease test re-enabled and passing

johanandren committed:
Verify lease error logs as well

johanandren committed:
Re-enabled PersistentShardSpec (and renamed)

johanandren committed:
WithLogCapturing where applicable

johanandren committed:
MiMa filters

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala:
```diff
@@ -557,6 +566,39 @@ private[akka] class ShardRegion(
   val initRegistrationDelay: FiniteDuration = 100.millis.max(retryInterval / 2 / 2 / 2)
   var nextRegistrationDelay: FiniteDuration = initRegistrationDelay
 
+  // FIXME not too happy about giving the shard actor context to the factory but the persistent one needs to spawn a child
```
This is the thing I was talking about @patriknw 

Unhappy with (but can't come up with a better strategy for) accessing the shard ActorContext from a lambda elsewhere (needed to spawn persistent per shard as child), but needs to be defined here to allow for a single shared instance for the ddata one.

Also, the magic reflection one only needed for testing is not nice. Sharding starting through an extension and not spawning the individual actors manually since they expect certain interaction and that would make for really brittle tests (there were a few I changed away from doing that)

patriknw replied:
does it have to be a child? would an alternative be to pass in system and it can create a systemActorOf?

patriknw replied:
or another level of indirection solves most problems ;-)

introducing a RememberEntitiesShardStoreProvider

```
trait RememberEntitiesShardStoreProvider {
  def actorResource: Option[Props]
  def createStore(factory: Option[ActorRef] => RememberEntitiesShardStore
```

then the actor would be created based on the `actorResource: Props`, and pass in that ref to the factory.

johanandren replied:
Hmm, yeah, maybe something like that is the path to go. I'll give it a shot.

johanandren committed:
Remove the old ddata shard impl (left it for comparing)

johanandren committed:
Base ask timeout on majority read/write timeouts

johanandren committed:
DData fixmes fixed

johanandren committed:
Timeouts from config for persistent remember entities

johanandren committed:
Single method for deliver, less utf-8 encoding

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala:
```diff
@@ -557,6 +566,39 @@ private[akka] class ShardRegion(
+          case ClusterShardingSettings.StateStoreModePersistence => { (shardId, shardCtx) =>
+            // one persistent store per shard
+            val store = shardCtx.actorOf(EventSourcedRememberEntitiesStore.props(typeName, shardId, settings))
+            context.watchWith(store, Shard.RememberEntityStoreCrashed(store)) // FIXME yuck
```
um, another `context`

johanandren committed:
Include detail on write failure

johanandren committed:
More structured providers, no access to region or shard contexts

johanandren commented:
Went with something inbetween the two suggestions:

 * introduced a provider api
 * changed the persistent actor to be spawned as a system actor

johanandren commented:
I think this is ready for more serious review, note that I have retargeted the PR to a new feature branch.

johanandren committed:
All tests green (woho!)

johanandren committed:
Don't send message to dead letter if it is actually handled in BackOffSupervisor

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -279,8 +242,76 @@ private[akka] class Shard(
+    case AkkaFailure(ex) =>
+      loadingEntityIdsFailed(ex)
+    case msg =>
+      log.debug("Got msg while waiting for remember entitites: {}", msg)
```
don't `toString` on unknown messages

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -279,8 +242,76 @@ private[akka] class Shard(
+      entityIds = ids
+      restartRememberedEntities(ids)
+    } else {
+      log.debug("Shard initialized")
```
it is initialized also when ids isEmpty so this log can go outside the if-else

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -432,594 +525,144 @@ private[akka] class Shard(
+   * @param entityIdWaitingForWrite an id for an remember entity write in progress, if non empty messages for that id
+   *                                will be buffered
+   */
+  private def deliverMessage(msg: Any, snd: ActorRef, entityIdWaitingForWrite: OptionVal[EntityId]): Unit = {
```
This is difficult, but also the area we want to explore performance optimizations for so more changes will probably come. 

Just some early thoughts for later. It can currently not start more than one entity at a time. That's how it was before also, but we could maybe have a store that can store for different entity ids in parallel. A single EventSourcedStore can't but the we could use more than one and hash/group on the entityId in similar way as we do for ddata.

johanandren replied:
One idea I had for improvement here would be to replace several lookups (messageBuffers, entityWaitingForWrite, refById, entityIds) with a single lookup, like a `statePerEntity: Map[EntityId, State]` where `State` could be something like `Idle(ref)`, `Passivating(ref)`, `WritInProgress(ref)` and perhaps a default `NotRunning` avoiding wrapping/unwrapping an `Option`.

Not so much for performance as for making it easier to follow.

chbatey replied:
Why can't we do it with a single event sourced store given we have one per shard? If the API took a AddEntities rather than AddEntity we could batch together multiple and persist them with a single persist.

At the moment we have the messages in a stash but we could have them in actor state and then when a write is complete to the remembered entities store send the rest as a batch?

I think this is a tradeoff between the simplicity of stashing and cost of doing one thing at a time.

patriknw reviewed (commented):
Looking good, but after seeing this I'm thinking that it would be better with an actor based spi instead of the future based.

I have reviewed in rather detail, but haven't looked much at the tests yet.

johanandren commented:
I agree on the actor vs future based SPI. Would be good with @chbatey s opinion as well before I start moving in that direction.

johanandren commented on akka-cluster-sharding/src/main/mima-filters/2.6.4.backwards.excludes/remember-entities-refactor.excludes:
```diff
@@ -0,0 +1,39 @@
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.sharding.Shard.this")
+
+# not marked internal but for not intended as public (no public API use case)
+ProblemFilters.exclude[MissingClassProblem]("akka.cluster.sharding.EntityRecoveryStrategy")
```
Yep, user api is strings in config/settings, not the actual classes.

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:
```diff
@@ -0,0 +1,220 @@
+
+    val store = system
+      .asInstanceOf[ExtendedActorSystem]
+      .systemActorOf(EventSourcedRememberEntitiesStore.props(typeName, shardId, settings), name)
```
Perhaps a good reasons to make sure this ends up as a child to the shard rather, since their lifecycles should be tied together.

patriknw replied:
yes, with actor based api we can have that, and other things like watch/supervise

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:
```diff
@@ -0,0 +1,220 @@
+  import EventSourcedRememberEntitiesStore._
+  import settings.tuningParameters._
+
+  log.debug("PersistentShard starting up [{}] [{}]", typeName, shardId)
```
Prefix perhaps not needed anymore at all since it is a separate actor/logger

johanandren committed:
Unprefix log entries

johanandren committed:
Back off supervisor log format plus use warning for hitting max restarts

johanandren committed:
Minor review feedback itemst addressed

johanandren commented on akka-actor/src/main/scala/akka/pattern/internal/BackoffOnStopSupervisor.scala:
```diff
@@ -61,11 +63,12 @@ import scala.concurrent.duration.FiniteDuration
         } else {
           log.debug(
-            s"Terminating on restart #{} which exceeds max allowed restarts ({})",
+            "Terminating on restart #{} which exceeds max allowed restarts ({})",
```
Changed, I also made it a warning because giving up on reaching a max restart number doesn't sound like debug to me.

johanandren committed:
Changed SPI to be actor/message based

One test failing

johanandren commented:
Rewrote the store using actors instead of a Future based SPI. One multi-jvm test failing that I need to look into.

patriknw commented:
actor based is looking good, left a few comments on that commit

chbatey commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -291,18 +332,75 @@ private[akka] class Shard(
+          case msg if extractEntityId.isDefinedAt(msg) => deliverMessage(msg, sender(), OptionVal.Some(entityId))
+          case msg                                     =>
+            // shouldn't be any other message types, but just in case
+            log.debug(
```
while on a branch put to warn or error so we see it?

chbatey commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -291,18 +332,75 @@ private[akka] class Shard(
+          case _: CoordinatorMessage                   => stash()
+          case _: ShardCommand                         => stash()
+          case _: ShardRegion.StartEntity              => stash()
+          case _: ShardRegion.StartEntityAck           => stash()
```
maybe while you're deep into this you can remind me why a shard received a start entity ack?

chbatey commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -291,18 +332,75 @@ private[akka] class Shard(
+
+        context.become {
+          case RememberEntitiesShardStore.UpdateDone(entityId) =>
+            if (VerboseDebug) log.debug("Completed async write of [{}] {}", entityId, command)
```
change log to be update done?

chbatey commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesStore.scala:
```diff
@@ -0,0 +1,196 @@
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object DDataRememberEntitiesStore {
```
👍

chbatey commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:
```diff
@@ -0,0 +1,156 @@
+/**
+ * INTERNAL API
+ *
+ * Persistent actor keeping the state for Akka Persistence backed remember entities (enabled through `state-store-mode=persistence`).
```
Pity this can't be implemented in typed. 
I guess as this is replacing a deprecated feature we can't remove yet we can't ask users to add a dep to typed sharding to get this

johanandren replied:
I feel the same :) 
And yes, I don't want to introduce the dependency (plus the cognitive load of having to keeping track of both classic and typed actor specifics in the same component)

chbatey commented on akka-cluster-sharding/src/test/scala/akka/cluster/sharding/RememberEntitiesFailureSpec.scala:
```diff
@@ -0,0 +1,290 @@
+    case class FailRemoveEntity(entityId: EntityId, whichWay: Fail)
+    case class DoNotFailRemoveEntity(entityId: EntityId)
+  }
+  class FakeStoreActor(shardId: ShardId) extends Actor with ActorLogging {
```
👍

chbatey reviewed (commented):
Looking good

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/RememberEntitiesStore.scala:
```diff
@@ -0,0 +1,43 @@
+ */
+@InternalApi
+private[akka] trait RememberEntitiesShardStoreProvider {
+  def shardStoreProps(shardId: ShardId): Props
```
In one iteration I decoupled that but then it gets messy with tying the lifecycles together and making sure the actor names are unique, solvable but much nicer like this.

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -291,18 +332,75 @@ private[akka] class Shard(
   private def receiveShardCommand(msg: ShardCommand): Unit = msg match {
-    // those are only used with remembering entities
+    // these are only used with remembering entities upon start
     case RestartEntity(id)    => getOrCreateEntity(id)
```
Yep, perhaps worth mentioning

johanandren commented:
I already have a note about batching up writes when a write is in progress that we can revisit when this refactor is done. I think that was the same as your idea @chbatey 👍

johanandren committed:
Some review feedback, test still failing

johanandren committed:
Missing assert that node had joined cluster

johanandren commented:
All tests passing now, and all outstanding feedback addressed, I think we can merge this into the feature branch now.

patriknw reviewed (approved):
LGTM, let's squash merge this to the feature branch
